csharp/src/Ice/UriParser.cs

```diff
@@ -0,0 +1,340 @@
+                    }
+                    if (oaEndpoints)
+                    {
+                        throw new FormatException("the option invocation-mode applies only to proxies");
+                    }
+                    if (int.TryParse(invocationModeValue, out int _))
+                    {
+                        throw new FormatException("the option invocation-mode does not accept numeric values");
+                    }
+                    invocationMode = Enum.Parse<InvocationMode>(invocationModeValue, ignoreCase: true);
+                }
+
+                string facet = uri.Fragment.Length >= 2 ? Uri.UnescapeDataString(uri.Fragment.Substring(1)) : "";
+                List<string> path = uri.AbsolutePath.Split('/').Select(s => Uri.UnescapeDataString(s)).ToList();
+
+                Debug.Assert(path.Count > 0); // there is always a first empty segment that we drop
+                Debug.Assert(path[0].Length == 0);
+                path.RemoveAt(0);
+
+                List<Endpoint>? endpoints = null;
+
+                if (endpointOptions != null) // i.e. not ice scheme
+                {
+                    endpoints = new List<Endpoint>
+                    {
+                        CreateEndpoint(communicator,
+                                       oaEndpoints,
+                                       endpointOptions,
+                                       protocol,
+                                       uri,
+                                       uriString)
+                    };
+
+                    if (generalOptions.TryGetValue("alt-endpoints", out string? altEndpointsValue))
+                    {
+                        string[] altEndpoints = altEndpointsValue.Split(',');
+                        foreach (string alt in altEndpoints)
+                        {
+                            if (alt.StartsWith("ice:"))
+                            {
+                                throw new FormatException(
+                                    $"invalid URI scheme for alt-endpoint `{alt}': must be empty or ice+transport");
+                            }
+
+                            string altUriString = alt;
+                            if (!altUriString.StartsWith("ice+"))
+                            {
+                                altUriString = $"{uri.Scheme}://{altUriString}";
+                            }
+                            altUriString = altUriString.Replace('$', '&');
+
+                            // No need to clear endpointOptions before reusing it since CreateEndpoint consumes all the
+                            // endpoint options
+                            Debug.Assert(endpointOptions.Count == 0);
+                            uri = InitialParse(altUriString, generalOptions: null, endpointOptions);
+
+                            // > 1 because there is always a first empty segment.
+                            if (uri.AbsolutePath.Length > 1 || uri.Fragment.Length > 0)
```

Comment:
Maybe a bit more readable to compare with `/`
```suggestion
                            if (uri.AbsolutePath != "/" || uri.Fragment.Length > 0)
```

Reply:
Added Assert instead.